generic_expression_patterns_modules/stats.py

```diff
@@ -0,0 +1,239 @@
+
+    # Need to temporarily write data to file in order
+    # to remove extra '\'
+    tmp_filename = os.path.join(local_dir, "pa_pathway_DB_tmp_filename.gmt")
+    pa_pathway_DB.to_csv(
+        tmp_filename,
+        quoting=csv.QUOTE_NONE,
+        escapechar="\\",
+        index=False,
+        header=False,
+        sep="\t",
+    )
+    with open(tmp_filename, "r") as ihf:
+        tmp = ihf.read()
+    with open(out_filename, "w") as ohf:
+        ohf.write(tmp.replace("\\", ""))
+
+
+def process_samples_for_limma(
+    expression_filename,
+    process_metadata_filename,
+    grp_metadata_filename,
+    out_expression_filename=None,
+):
+    """
+    This function processes samples in the template and simulated
+    experiments to prepare for DE analysis using DESeq.
+
+    These processing steps includes:
+    1. Removing samples that are not included in the comparison.
+    These "extra" samples occur when an experiment contains multiple
+    comparisons.
+    2. Checks that the ordering of samples in the metadata file
+    are consistent with the ordering in the gene expression data
+    matrix. If the ordering is not consistent, then samples in
+    the gene expression data matrix are re-ordered.
+
+    Arguments
+    ----------
+    expression_filename: str
+        File containing unnormalized gene expression data for
+        either template or simulated experiments
+    process_metadata_filename: str
+        File containing assignment for which samples to drop
+    grp_metadata_filename: str
+        File containing group assigments for samples to use
+        for DESeq analysis
+    out_expression_filename (optional): str
+        File to save processed gene expression data to.
+        If not provided processed gene expression data will
+        be output to the same input filename   
+
+    """
+
+    # Read data
+    expression = pd.read_csv(expression_filename, sep="\t", index_col=0, header=0)
+    process_metadata = pd.read_csv(
+        process_metadata_filename, sep="\t", index_col=0, header=0
+    )
+    grp_metadata = pd.read_csv(grp_metadata_filename, sep="\t", header=0, index_col=0)
+
+    # Get samples ids to remove
+    samples_to_remove = list(
+        process_metadata[process_metadata["processing"] == "drop"].index
+    )
+
+    # Remove samples
+    expression = expression.drop(samples_to_remove)
+
+    # Check ordering of sample ids is consistent between gene expression data and metadata
+    metadata_sample_ids = grp_metadata.index
+    expression_sample_ids = expression.index
+
+    if metadata_sample_ids.equals(expression_sample_ids):
+        print("sample ids are ordered correctly")
+    else:
+        # Convert gene expression ordering to be the same as
+        # metadata sample ordering
+        print("sample ids don't match, going to re-order gene expression samples")
+        expression = expression.reindex(metadata_sample_ids)
+
+    # Save
+    if out_expression_filename != None:
```

Comment:
When I'm comparing against None, I typically try to use `is not None` rather than `!= None`, although in most cases it doesn't matter. (Also, `is not None` is slightly faster)

See here for some cases when it does matter and an explanation: http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html .